src/lib/components/Filter.svelte

```diff
@@ -0,0 +1,234 @@
+
+  $: typeList = _.chain(data)
+    .flatMap('type')
+    .uniq()
+    .value()
+    .map((value, i) => ({
+      value: value,
+      index: i,
+      label: value
+        .split('_')
+        .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
+        .join(' '),
+    }));
+
+  function fromParam(param, lst, complex) {
+    const value_ = $page.url.searchParams.get(param);
+    if (complex === true) {
+      lst = _.chain(lst).flatMap('value').value();
+    }
+    if (lst.includes(value_)) {
+      return {
+        value: value_,
+        label: value_
+          .split('_')
+          .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
+          .join(' '),
+        index: lst.indexOf(value_),
+      };
+    }
+  }
+
+  function onChange() {
+    if ($page.url.searchParams.get('lc')) {
+      chapter = fromParam('lc', chapterList);
+    }
+    if ($page.url.searchParams.get('type')) {
+      type = fromParam('type', typeList, true);
+    }
+    if ($page.url.searchParams.get('lang')) {
+      lang = fromParam('lang', langList, true);
+    }
+    if ($page.url.searchParams.get('tag')) {
+      console.log(tag);
+      const tag_ = $page.url.searchParams.get('tag').split(',');
+      const arr = [];
+      for (let i = 0; i < tag_.length; i++) {
+        if (tagList.includes(tag_[i])) {
+          arr.push({
+            value: tag_[i],
+            label: tag_[i],
+            index: tagList.indexOf(tag_[i]),
+          });
+        }
+      }
+      tag = arr;
+    }
+  }
+
+  $: onChange($page.url.searchParams);
+
+  function filter(data_, type, lang, tag, chapter) {
+    if (chapterList) {
+      if (chapter) {
+        data_ = filterByChapter(data_, chapter.value);
+      } else {
+      }
+    }
+    if (langList) {
+      if (lang) {
+        data_ = filterBy('language', data_, lang.value);
+      } else {
+      }
+    }
+
+    if (tag) {
+      tag = _.chain(tag).flatMap('value').value();
+      data_ = filterByTag(data_, tag);
+    } else {
+    }
+
+    if (type) {
+      data_ = filterBy('type', data_, type.value);
+    } else {
+    }
+
+    return data_;
+  }
+
+  $: $filter_data = filter(data, type, lang, tag, chapter);
+  $: {
+    const newUrl = new URL($page.url);
+
+    if (chapter) {
+      newUrl?.searchParams?.set('lc', chapter.value);
+    } else {
+      newUrl?.searchParams?.delete('lc');
+    }
+    if (type) {
+      newUrl?.searchParams?.set('type', type.value);
+    } else {
+      newUrl?.searchParams?.delete('type');
+    }
+    if (tag) {
+      newUrl?.searchParams?.set('tag', _.chain(tag).flatMap('value').value());
+    } else {
+      newUrl?.searchParams?.delete('tag');
+    }
+    if (lang) {
+      newUrl?.searchParams?.set('lang', lang.value);
+    } else {
+      newUrl?.searchParams?.delete('lang');
+    }
+
+    goto(newUrl);
+  }
+
+  let hidden = 'hidden';
+  $: if (chapter || type || lang || tag) {
+    hidden = 'visible';
+  }
+  function handle_hidden() {
```

Comment:
```suggestion
  function handleHidden() {
```

It would probably be even better to rename it wrt. what it does. For instance `toggleVisibility`.

Reply:
alright adjusted this